### PR TITLE
Remove dynamic alloc of PROBLEME_HEBDO

### PR DESCRIPTION
**Pull request by payetvin**

### CODE REVIEW

File: src/solver/simulation/adequacy.cpp [hunk 1]
```diff
@@ -35,27 +35,25 @@ using Antares::Constants::nbHoursInAWeek;
...
 {
-Adequacy::Adequacy(Data::Study& study) : study(study), preproOnly(false), pProblemesHebdo(nullptr)
+Adequacy::Adequacy(Data::Study& study) : study(study), preproOnly(false)
 {
 }
 
 Adequacy::~Adequacy()
 {
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
```
Comment (sylvlecl) [inline src/solver/simulation/adequacy.cpp]: Simpler to remove this unnecessary check

File: src/solver/simulation/adequacy.cpp [hunk 2]
```diff
@@ -80,26 +78,25 @@ bool Adequacy::simulationBegin()
...
+            if ((uint)nbHoursInAWeek != (uint)pProblemesHebdo[numSpace].NombreDePasDeTemps)
             {
                 logs.fatal() << "internal error";
                 return false;
             }
         }
     }
 
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
```
Comment (sylvlecl) [inline src/solver/simulation/adequacy.cpp]: we can remove it

File: src/solver/optimisation/opt_alloc_probleme_a_optimiser.cpp [hunk 3]
```diff
@@ -176,9 +176,9 @@ void OPT_AugmenterLaTailleDeLaMatriceDesContraintes(PROBLEME_ANTARES_A_RESOUDRE*
     ProblemeAResoudre->NombreDeTermesAllouesDansLaMatriceDesContraintes = NbTermes;
 }
 
-void OPT_LiberationMemoireDuProblemeAOptimiser(PROBLEME_HEBDO* problemeHebdo)
+void OPT_LiberationMemoireDuProblemeAOptimiser(PROBLEME_HEBDO& problemeHebdo)
 {
-    PROBLEME_ANTARES_A_RESOUDRE* ProblemeAResoudre = problemeHebdo->ProblemeAResoudre;
+    PROBLEME_ANTARES_A_RESOUDRE* ProblemeAResoudre = problemeHebdo.ProblemeAResoudre;
```
Comment (flomnes) [inline src/solver/optimisation/opt_alloc_probleme_a_optimiser.cpp]: From my understanding, `ProblemeAResoudre` is now a `unique_ptr`(see #1549). I recommend removing function `OPT_LiberationMemoireDuProblemeAOptimiser`.

File: src/solver/simulation/economy.cpp [hunk 4]
```diff
@@ -39,27 +39,25 @@ using Antares::Constants::nbHoursInAWeek;
...
 {
-Economy::Economy(Data::Study& study) : study(study), preproOnly(false), pProblemesHebdo(nullptr)
+Economy::Economy(Data::Study& study) : study(study), preproOnly(false)
 {
 }
 
 Economy::~Economy()
 {
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
```
Comment (sylvlecl) [inline src/solver/simulation/economy.cpp]: we can remove this

File: src/solver/simulation/adequacy.cpp (hunk 1 above)
Comment (flomnes) [inline src/solver/simulation/adequacy.cpp]: No longer required

File: src/solver/simulation/economy.cpp [hunk 5]
```diff
@@ -118,10 +115,10 @@ bool Economy::simulationBegin()
         }
     }
 
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
```
Comment (sylvlecl) [inline src/solver/simulation/economy.cpp]: to be removed

File: src/solver/utils/optimization_statistics.h [hunk 6]
```diff
@@ -28,6 +28,14 @@ class OptimizationStatistics
         this->reset();
     }
 
+    OptimizationStatistics(const OptimizationStatistics &o)
```
Comment (flomnes) [inline src/solver/utils/optimization_statistics.h]:
```suggestion
    OptimizationStatistics(const OptimizationStatistics &rhs)
```

**Review by flomnes — CHANGES_REQUESTED** [review]

File: src/solver/simulation/sim_structure_probleme_economique.h [hunk 7]
```diff
@@ -478,6 +478,11 @@ struct VARIABLES_DUALES_INTERCONNEXIONS
 
 struct PROBLEME_HEBDO
 {
+    PROBLEME_HEBDO() = default;
+    PROBLEME_HEBDO(const PROBLEME_HEBDO&) = delete;
+    PROBLEME_HEBDO(PROBLEME_HEBDO&&) = default;
+    PROBLEME_HEBDO& operator=(PROBLEME_HEBDO&& other) = default;
```
Comment (sylvlecl) [inline src/solver/simulation/sim_structure_probleme_economique.h]: See "[rule of 5](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all)": let's define also explicitly the = operator as deleted, for clarity sake.

File: src/solver/utils/optimization_statistics.h (hunk 6 above)
Comment (sylvlecl) [inline src/solver/utils/optimization_statistics.h]: See "[rule of 5](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all)": for clarity sake,
let's define also copy operator and move operators (or delete them).

File: src/solver/simulation/adequacy.cpp [hunk 8]
```diff
@@ -80,26 +78,25 @@ bool Adequacy::simulationBegin()
...
     {
         for (uint numSpace = 0; numSpace < pNbMaxPerformedYearsInParallel; numSpace++)
```
Comment (sylvlecl) [inline src/solver/simulation/adequacy.cpp]: we can now just iterate on the vector content, for example:
```cpp
for (auto& problem: pProblemesHebdo)
{
    pProblemesHebdo.TypeDOptimisation = OPTIMISATION_LINEAIRE;
}
```

File: src/solver/simulation/adequacy.cpp [hunk 9]
```diff
@@ -35,27 +35,25 @@ using Antares::Constants::nbHoursInAWeek;
...
+Adequacy::Adequacy(Data::Study& study) : study(study), preproOnly(false)
 {
 }
 
 Adequacy::~Adequacy()
 {
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
     {
         for (uint numSpace = 0; numSpace < pNbMaxPerformedYearsInParallel; numSpace++)
```
Comment (sylvlecl) [inline src/solver/simulation/adequacy.cpp]: We can now just iterate on the vector content.

File: src/solver/simulation/economy.cpp [hunk 10]
```diff
@@ -39,27 +39,25 @@ using Antares::Constants::nbHoursInAWeek;
...
+Economy::Economy(Data::Study& study) : study(study), preproOnly(false)
 {
 }
 
 Economy::~Economy()
 {
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
     {
         for (uint numSpace = 0; numSpace < pNbMaxPerformedYearsInParallel; numSpace++)
```
Comment (sylvlecl) [inline src/solver/simulation/economy.cpp]: We can now just iterate on the vector content.

File: src/solver/simulation/economy.cpp [hunk 11]
```diff
@@ -118,10 +115,10 @@ bool Economy::simulationBegin()
         }
     }
 
-    if (pProblemesHebdo)
+    if (!pProblemesHebdo.empty())
     {
         for (uint numSpace = 0; numSpace < pNbMaxPerformedYearsInParallel; numSpace++)
```
Comment (sylvlecl) [inline src/solver/simulation/economy.cpp]: We can now just iterate on the vector content.

**Review by sylvlecl — CHANGES_REQUESTED** [review]
A few more improvements are possible now that we use the vector.
Also, let's try to follow "rule of 3/5".

**Review by flomnes — CHANGES_REQUESTED** [review]
The tests are failing on Windows, please investigate.

**Review by flomnes — CHANGES_REQUESTED** [review]
Add TODO note in economy.cpp / adequacy.cpp, about why we still need a "classic" loop to iterate on a `std::vector`.

**Review by flomnes — APPROVED** [review]

**Review by sylvlecl — APPROVED** [review]

**Pull request merged**